wherehows-web/app/components/browser/containers/browser-viewport.ts

```diff
@@ -36,34 +41,110 @@ export default class BrowserViewport extends Component {
-   * @type {TaskProperty<Promise<IDatasetView[]>> & {perform: (a?: {} | undefined) => TaskInstance<Promise<IDatasetView[]>>}}
+   * Async request a list of datasets with a variable start point
+   * @type {TaskProperty<Promise<IDatasetsGetResponse>> & {perform: (a1: "push" | "set" | undefined, a2: number) => TaskInstance<Promise<IDatasetsGetResponse>>}}
    * @memberof BrowserViewport
    */
-  getEntitiesTask = task(function*(this: BrowserViewport): IterableIterator<Promise<Array<IDatasetView>>> {
+  getDatasetsTask = task(function*(
+    this: BrowserViewport,
+    op: listOp = 'set',
+    start: number
+  ): IterableIterator<Promise<IDatasetsGetResponse>> {
     const { prefix, platform, entity } = get(this, 'params');
+    const response = yield readDatasets({ platform, prefix, start });
+    const { total, count, elements } = response;
+    const entities = get(this, 'entities');
+    // If new segment / datasource, replace all items rather than append
+    const listOp = op === 'set' ? [].setObjects : [].pushObjects;
 
-    const entities = (<IGetEntityTaskStrategy>{
-      datasets: readDatasets({ platform, prefix }),
-      metrics: Promise.resolve([]),
-      flows: Promise.resolve([])
-    })[entity];
+    listOp.call(entities, elements);
 
     setProperties(this, {
-      entities: yield entities,
+      total,
+      count,
       currentEntity: entity,
-      entityRoute: `${entity}.${entity.slice(0, -1)}`
+      entityRoute: `${entity}.${entity.slice(0, -1)}`,
+      start: entities.length
     });
+
+    return response;
+  });
+
+  /**
+   * Async requests for the list of entities and sets the value on class
+   * @type {TaskProperty<Promise<IDatasetsGetResponse>> & {perform: (a?: number) => TaskInstance<Promise<IDatasetsGetResponse>>}}
+   * @memberof BrowserViewport
+   */
+  getEntityDataTask = task(function*(
+    this: BrowserViewport,
+    offset: number = 0
+  ): IterableIterator<TaskInstance<Promise<IDatasetsGetResponse>>> {
+    const { entity } = get(this, 'params');
+
+    yield (<IGetEntityTaskStrategy>{
+      datasets: get(this, 'getDatasetsTask').perform('set', offset)
+    })[entity];
+  });
+
+  /**
+   * The max possible entities that can currently be requested up to this.count
+   * @type {ComputedProperty<number>}
+   * @memberof BrowserViewport
+   */
+  nextCountSize: ComputedProperty<number> = computed('entities.length', function(this: BrowserViewport): number {
+    const { entities: { length }, total, count } = getProperties(this, ['entities', 'total', 'count']);
+    return Math.min(count, total - length);
+  });
+
+  /**
+   * The total number of items left that can be fetched
+   * @type {ComputedProperty<number>}
+   * @memberof BrowserViewport
+   */
+  remainingEntityCount: ComputedProperty<number> = computed('entities.length', function(this: BrowserViewport): number {
+    const { entities: { length }, total } = getProperties(this, ['entities', 'total']);
```

Comment:
you can do getProperties(this, 'entities.length')

Reply:
@igbopie not in typescript <=2.7, typescript limitation with the ember object model :/
https://github.com/typed-ember/ember-cli-typescript#nested-keys-in-get-or-set